### DISCUSSION TITLE
thenable.then is not called syncly?

### DISCUSSION THREAD

**Discussion by Worldwidebrine** [mdn/mdn-community#523] [Content] [answered]
```js
console.log(1, "returned promise", Promise.resolve({
    then(broadcastFulfillment, broadcastRejection) {
        console.log(1, "called .then");
    },
    log: (() => console.log(1, "thenable object evaluated"))()
}));
console.log(2, "returned promise", Promise.resolve(Promise.resolve()));
```

```
1 'thenable object evaluated'
1 'returned promise' Promise {<pending>}
2 'returned promise' Promise {<fulfilled>: undefined}
1 'called .then'
```

**Comment by Josh-Cena** [comment, marked as answer]
Yep, you can see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise#resolver_function:

> If it's called with a thenable value (including another `Promise` instance), then the thenable's `then` method is saved and called in the future (it's always called asynchronously). The then method will be called with two callbacks, which are two new functions with the exact same behaviors as the `resolveFunc` and `rejectFunc` passed to the `executor` function. If calling the `then` method throws, then the current promise is rejected with the thrown error.

The `Promise.resolve` page mentions that for non-native promises, it's essentially a shorthand for `new Promise((res) => res(value))`, and that

> The bulk of the resolving logic is actually implemented by the [resolver function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise#resolver_function) passed by the `Promise()` constructor.

**Reply by Worldwidebrine** [reply]
Thanks. I forgot thenable object is decoupled from native promise, so I thought thenable needs to pass onres and onrej to the native promise side, that was completely wrong.
